private/buf/bufcli/bufcli.go

```diff
@@ -725,6 +739,93 @@ func ReadModuleWithWorkspacesDisabled(
 	return module, moduleIdentity, err
 }
 
+// NewImageForSource resolves a single bufimage.Image from the user-provided source with the build options.
+func NewImageForSource(
+	ctx context.Context,
+	container appflag.Container,
+	source string,
+	errorFormat string,
+	disableSymlinks bool,
+	configOverride string,
+	externalDirOrFilePaths []string,
+	externalExcludeDirOrFilePaths []string,
+	externalDirOrFilePathsAllowNotExist bool,
+	excludeSourceCodeInfo bool,
+) (bufimage.Image, error) {
+	ref, err := buffetch.NewRefParser(container.Logger(), buffetch.RefParserWithProtoFileRefAllowed()).GetRef(ctx, source)
+	if err != nil {
+		return nil, err
+	}
+	storageosProvider := NewStorageosProvider(disableSymlinks)
+	runner := command.NewRunner()
+	registryProvider, err := NewRegistryProvider(ctx, container)
+	if err != nil {
+		return nil, err
+	}
+	imageConfigReader, err := NewWireImageConfigReader(
+		container,
+		storageosProvider,
+		runner,
+		registryProvider,
+	)
+	if err != nil {
+		return nil, err
+	}
+	imageConfigs, fileAnnotations, err := imageConfigReader.GetImageConfigs(
+		ctx,
+		container,
+		ref,
+		configOverride,
+		externalDirOrFilePaths,
+		externalExcludeDirOrFilePaths,
+		externalDirOrFilePathsAllowNotExist,
+		excludeSourceCodeInfo,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(fileAnnotations) > 0 {
+		// stderr since we do output to stdout potentially
+		if err := bufanalysis.PrintFileAnnotations(
+			container.Stderr(),
+			fileAnnotations,
+			errorFormat,
+		); err != nil {
+			return nil, err
+		}
+		return nil, ErrFileAnnotation
+	}
+	images := make([]bufimage.Image, 0, len(imageConfigs))
+	for _, imageConfig := range imageConfigs {
+		images = append(images, imageConfig.Image())
+	}
+	return bufimage.MergeImages(images...)
+}
+
+// ParseSourceAndType returns the moduleReference and typeName from the source and type provided by the user.
+// When source is not provided, we assume the type is a fully-qualified path to the type and try to parse it.
+// Otherwise, if both source and type are provided, the type must be a valid Protobuf identifier (e.g. weather.v1.Units).
+func ParseSourceAndType(
+	ctx context.Context,
+	flagSource string,
+	flagType string,
+) (moduleReference string, typeName string, _ error) {
```

Comment:
We might not want to use name arguments here because we're not always returning a `moduleReference`. The `flagSource` could be a local directory, and it could be returned as-is if both `--source` and `--type` are provided.

I'd just drop the return argument names. Also, I'd drop the `flag` prefix from `flagSource` and `flagType` and potentially relabel `type` to `typeName` to avoid the keyword conflict.